lib/remote/jsonrpcconnection-pki.cpp

```diff
@@ -61,7 +64,7 @@ Value RequestCertificateHandler(const MessageOrigin::Ptr& origin, const Dictiona
 		logmsg << "Received certificate request for CN '" << cn << "'";
 
 		try {
-			signedByCA = VerifyCertificate(cacert, cert, listener->GetCrlPath());
+			signedByCA = VerifyCertificate(cacert, cert, listener->GetCrlPath(), chain);
```

Comment:
... make pki::RequestCertificate recognise the requestor as legit cluster member.

But **only pki::RequestCertificate**. Which will issue you a direct leaf cert (root -> leaf) as replacement for your indirect one (root -> int. -> leaf) if the root CA key is on the file system. Or an indirect one if you put the int. CA as root into /var/lib/icinga2/ca. Anyway.

I'd not change that, but I don’t see how this alone makes Icinga 2 accepting leaf as in _root -> int. -> leaf_ as legit cluster member. Please explain.

Reply:
> * the master has to have `root` in its `ca.pem` (either only the root

> * the agent has its `leaf` in its `cert.pem` and the `intermediate` either following the leaf in the same file,

So again you say that https://github.com/Icinga/icinga2/pull/8859#issuecomment-1551364930 should just work. Demonstrate it. Fresh cluster w/o CA **key**, external root and intermediate CA, the latter ONLY in the same files as the leaves. What does the Icinga 2 log say about established connections?

Reply:
Sure. So I downloaded a clean tarball of 2.13.7, added the commit from this PR and made a fresh build. Setup master and slave without any CA or certs from icinga itself.

For certificates I used some test-certificates from our puppet environment. Setup as follows:
* icinga-master (called `rocky9eval.XXX.net`), only having the CA-root.crt, no intermediate at all (the `ca.crt` only has the root):
```
~~~icinga-devel~~id-master~~openssl x509 -text -in var/lib/icinga2/certs/ca.crt
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
        Issuer: CN = Puppet Root CA: 9bf8c56a6ec228
        Subject: CN = Puppet Root CA: 9bf8c56a6ec228
```

* icinga-agent (called `rsdevel.XXX.net`), also only having the CA-root.crt and no intermediate at all (also `ca.crt` only has the root, and `rsdevel.XXX.net.crt` only the leaf):
```
~~~icinga-devel~~id-agent~~openssl x509 -text -in var/lib/icinga2/certs/ca.crt
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
        Issuer: CN = Puppet Root CA: 9bf8c56a6ec228
        Subject: CN = Puppet Root CA: 9bf8c56a6ec228
~~~icinga-devel~~id-agent~~openssl x509 -text -in var/lib/icinga2/certs/rsdevel.XXX.net.crt
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 8 (0x8)
        Issuer: CN = "Puppet Enterprise CA generated at +2021-02-15 18:26:49 +0100"
        Subject: CN = rsdevel.XXX.net
```

Now starting the icinga2-daemon on the agent, and then on the master (which then will try to connect to the agent), resulting in:
```
~~~icinga-devel~~id-master~~./sbin/icinga2 daemon·
[...]
[2023-06-21 13:59:02 +0200] warning/ApiListener: Certificate validation failed for endpoint 'rsdevel.XXX.net': code 21: unable to verify the first certificate
[2023-06-21 13:59:02 +0200] information/ApiListener: New client connection for identity 'rsdevel.XXX.net' to [::1]:15665 (certificate validation failed: code 21: unable to verify the first certificate)
[2023-06-21 13:59:02 +0200] information/ApiListener: Finished reconnecting to endpoint 'rsdevel.XXX.net' via host 'localhost' and port '15665'
```

So far as expected.

Now I add the intermediate to the leaf-certificate on the agent-side:
```
~~~icinga-devel~~id-agent~~openssl x509 -text -in var/lib/icinga2/certs/ca-int.pem·
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 2 (0x2)
        Issuer: CN = Puppet Root CA: 9bf8c56a6ec228
        Subject: CN = "Puppet Enterprise CA generated at +2021-02-15 18:26:49 +0100"
~~~icinga-devel~~id-agent~~cat var/lib/icinga2/certs/ca-int.pem >> var/lib/icinga2/certs/rsdevel.XXX.net.crt
```

Restart the daemon on the agent- and master-side. Now the master is able to connect:
```
~~~icinga-devel~~id-master~~./sbin/icinga2 daemon·
[...]
[2023-06-21 14:06:58 +0200] information/ApiListener: New client connection for identity 'rsdevel.XXX.net' to [::1]:15665
[2023-06-21 14:06:58 +0200] information/ApiListener: Sending config updates for endpoint 'rsdevel.XXX.net' in zone 'rsdevel.XXX.net'.
```

Reply:
I guess this works because one successful VerifyCertificate() is enough. I bet if you comment out sending pki::RequestCertificate, it stops working. Please verify.

Reply:
> I guess this works because one successful VerifyCertificate() is enough. I bet if you comment out sending pki::RequestCertificate, it stops working. Please verify.

Hmmm, I commented all four occurences of `JsonRpcConnection::SendCertificateRequest` (three in `lib/remote/apilistener.cpp` and one in `lib/remote/jsonrpcconnection-pki.cpp`) and recompiled for master and agent side, and everything still works. But actually I don't understand why you think that this would break the setup.

As far as I understand, the connection is established in `lib/remote/apilistener.cpp` via `ApiListener::UpdateSSLContext()`. The function takes the filename of the nodes-certificate-file (which in this case includes the leaf+intermediate) and hands that filename down the whole SSL-chain to `InitSslContext()` in `lib/base/tlsutility.cpp`. There that file is finally given to OpenSSL by calling `SSL_CTX_use_certificate_chain_file()`.

According to the manual page:
```
       SSL_CTX_use_certificate_chain_file() loads a certificate chain from file into ctx. The certificates must be in PEM format and must be sorted starting with the subject's certificate (actual client or server certificate), followed by intermediate CA certificates if applicable, and ending at the highest
       level (root) CA. SSL_use_certificate_chain_file() is similar except it loads the certificate chain into ssl.
```

So regardless what, the agent will always put its whole content of the local `cert.pem` on the wire (even with a totally unpatched icinga-version). The first `VerifyCertificate()` that is called, is just the one which was patched here, giving the chain from the wire to the verification as well. As far as I see, the only other `VerifyCertificate()`-call that exists (outside the CLI), is the one in the `RenewCert()` function, which shouldn't be called anyway, since the internal icinga-CA isn't used here.

Reply:
Then it likely works because Boost ASIO "just does it".

Reply:
Well, more like OpenSSL already does it, but yes.

If you start and agent with an unpatched 2.13.7 and just connect to the API-port, you see that the whole content of the `cert.pem` goes onto the wire:
```
~~~openssl s_client -connect localhost:15665
CONNECTED(00000003)
Can't use SSL_get_servername
depth=1 CN = "Puppet Enterprise CA generated at +2021-02-15 18:26:49 +0100"
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = rsdevel.XXX.net
verify return:1
---
Certificate chain
 0 s:CN = rsdevel.XXX.net
   i:CN = "Puppet Enterprise CA generated at +2021-02-15 18:26:49 +0100"
   a:PKEY: rsaEncryption, 4096 (bit); sigalg: RSA-SHA256
   v:NotBefore: Jun 27 16:25:02 2021 GMT; NotAfter: Jun 27 16:25:02 2026 GMT
 1 s:CN = "Puppet Enterprise CA generated at +2021-02-15 18:26:49 +0100"
   i:CN = Puppet Root CA: 9bf8c56a6ec228
   a:PKEY: rsaEncryption, 4096 (bit); sigalg: RSA-SHA256
   v:NotBefore: Feb 14 17:27:20 2021 GMT; NotAfter: Feb 12 17:27:26 2036 GMT
---
```